pkg/validate/container_service.go

```diff
@@ -1,10 +1,258 @@
+		errs = append(errs, fmt.Errorf("invalid properties.servicePrincipalProfile.secret %q", spp.Secret))
+	}
+
+	return
+}
+
+func validateAgentPoolProfiles(apps []*api.AgentPoolProfile) (errs []error) {
+	appmap := map[string]*api.AgentPoolProfile{}
+
+	for i, app := range apps {
+		// TODO why is this a pointer?
+		if app == nil {
+			continue
+		}
+
+		if _, found := validAgentPoolProfileNames[app.Name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", app.Name))
+		}
+
+		if _, found := appmap[app.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles %q", app.Name))
+		}
+		appmap[app.Name] = app
+
+		if i > 0 && app.VnetSubnetID != apps[i-1].VnetSubnetID {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles.vnetSubnetID %q", app.VnetSubnetID))
+		}
+
+		errs = append(errs, validateAgentPoolProfile(app)...)
+	}
+
+	for name := range validAgentPoolProfileNames {
+		if _, found := appmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", name))
+		}
+	}
+
+	return
+}
+
+func validateAgentPoolProfile(app *api.AgentPoolProfile) (errs []error) {
+	if app.Name != string(app.Role) {
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].name %q", app.Name, app.Name))
+	}
+
+	switch app.Role {
+	case api.AgentPoolProfileRoleCompute,
+		api.AgentPoolProfileRoleInfra:
+		if app.Count < 1 || app.Count > 100 {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].count %q", app.Name, app.Count))
+		}
+
+	case api.AgentPoolProfileRoleMaster:
+		if app.Count != 3 {
+			errs = append(errs, fmt.Errorf("invalid masterPoolProfile.count %d", app.Count))
+		}
+
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].role %q", app.Name, app.Role))
+	}
+
+	switch app.VMSize {
+	case "Standard_D2s_v3",
+		"Standard_D4s_v3":
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].vmSize %q", app.Name, app.VMSize))
+	}
+
+	if app.VnetSubnetID != "" && !rxAgentPoolProfileVNetSubnetID.MatchString(app.VnetSubnetID) {
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].vnetSubnetID %q", app.Name, app.VnetSubnetID))
+	}
+
+	switch app.OSType {
+	case api.OSTypeLinux:
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].osType %q", app.Name, app.OSType))
+	}
+
+	return
+}
+
+func validateOrchestratorProfile(p *api.OrchestratorProfile) (errs []error) {
+	if p == nil {
+		errs = append(errs, fmt.Errorf("orchestratorProfile cannot be nil"))
+		return
+	}
+
+	switch p.OrchestratorVersion {
+	case "v3.10":
+	default:
+		errs = append(errs, fmt.Errorf("invalid properties.openShiftVersion %q", p.OrchestratorVersion))
+	}
+
+	errs = append(errs, validateOpenShiftConfig(p.OpenShiftConfig)...)
+
+	return
+}
+
+func validateMasterProfile(p *api.MasterProfile) (errs []error) {
```

Comment:
Should this have validation for `count` as well (3 or 5 as allowed values)?

Reply:
no, master is an agent pool in the internal representation so that validation occurs here: https://github.com/openshift/openshift-azure/pull/172/commits/17ac105e64003d35eaa5a0cd2acd182160832949#diff-33ccf5fc1c1988bc397bbbabb210cde4R134